Add documentation about podvm build process

wainersm opened the pull request:
While working on https://github.com/confidential-containers/cloud-api-adaptor/pull/345 we realized the process to build podvm within container is not documented, can be confusing and demands memory/disk from the host. So the main goal of this PR is to document the process to build the podvm with or without containers.

And while working on the podvm/README.md  I realized that:

 * we don't document how to consume the pre-built images either, so I add a new article in docs.
 * we could have a script to ease the process to extract the qcow2 file from the container image, so I created a script. A few minutes about I noticed @jtumber-ibm  had the same idea (https://github.com/confidential-containers/cloud-api-adaptor/issues/563) and his script (https://github.com/confidential-containers/cloud-api-adaptor/pull/565/commits/de6725588e474ffeb9853d08fec53b7ba6244e76) is better actually :)

I didn't document the tricks to build the RHEL image because I never built it and I actually I lack time to try. Maybe @snir911 and/or @whitebrandy could help here? :)

wainersm commented:
hi @jtumber-ibm ,

I tried to build the s390x image on my x86_64 laptop just like I documented on the podvm/README.md but it failed on the packer step. Did I miss something?

Also, I am okay to drop my extract-qcow2.sh in favor of your script. I will just need to ensure it is merged before and I can change the docs.

snir911 commented on docs/consuming-prebuilt-podvm-images.md:
````diff
@@ -0,0 +1,14 @@
+$ ./hack/extract-qcow2.sh quay.io/confidential-containers/podvm-libvirt-ubuntu
+```
+
+If you need to customize the image, for example, to add a [registries authentication configuration](registries-authentication.md) then it can be done with the [libguestfs](https://www.libguestfs.org) tools. Alternatively, you can find instructions on [podvm/README.md](../podvm/README.md) on how to build the image from sources.
````
can you use [statically embedded registries authentication configuration](registries-authentication.md#statically-embed-authentication-file-in-podvm-image) instead? as it's not needed if auth file passed at runtime
raw:
`[statically embedded registries authentication configuration](registries-authentication.md#statically-embed-authentication-file-in-podvm-image)`

wainersm replied:
I am still a little confused here but I gave it a try and re-wrote that section. Let me know if it looks good (or not).

snir911 commented on podvm/README.md:
```diff
@@ -0,0 +1,128 @@
+the instructions on the next sections if you need to build your own image with changes to meet your requirements. Otherwise you can
+find [here](../docs/consuming-prebuilt-podvm-images.md) information on how to consume pre-built images.
+
+# How to build on locally
```
remove "on"

snir911 commented on podvm/README.md:
```diff
@@ -0,0 +1,128 @@
+
+# How to build on locally
+
+In order to build locally you must [packer](https://www.packer.io/) and QEMU installed in your host. Apart from those, it requires the source trees and software mentioned on the [developer's guide](docs/DEVELOPMENT.md).
```
you must **have** ?

snir911 commented on podvm/README.md:
````diff
@@ -0,0 +1,128 @@
+```
+$ export CLOUD_PROVIDER=[aws|azure|ibmcloud|libvirt]
+$ make binaries
+$ make image
````
nit: make image would be enough

wainersm replied:
ah, cool. I will adjust that.

jtumber-ibm commented:
> hi @jtumber-ibm ,
> 
> I tried to build the s390x image on my x86_64 laptop just like I documented on the podvm/README.md but it failed on the packer step. Did I miss something?
> 
> Also, I am okay to drop my extract-qcow2.sh in favor of your script. I will just need to ensure it is merged before and I can change the docs.

hi @wainersm, did you have the s390x components installed? i.e. rust target (s390x-unknown-linux-gnu), qemu (qemu-system-s390x) and gcc (gcc-s390x-linux-gnu).

Feel free to use my script, I just need a couple other IBMers to test the ibmcloud part of the script but should be able to get that merged soon or if that gets delayed I can create a separate PR for the qcow2 from image bit.

snir911 commented on podvm/README.md:
```diff
@@ -0,0 +1,128 @@
+
+In general it is needed to follow these steps:
+
+1. Build a builder container image for a given Linux distribution
```
Ideally we can have most builders publicly available, but i think we didn't push them

bpradipt replied:
Yeah that was the intent. Having podvm_builder images published for releases (including any pre-releases)

wainersm replied:
Agreed. Actually I created this README aiming at developers that need to build podvm from latest code. And the [docs/consuming-prebuilt-podvm-images.md](https://github.com/confidential-containers/cloud-api-adaptor/pull/568/files/d5dab40a507fcd557cfa9747547ee6014a77dc78#diff-4fc2fb53a97f8dc41a360dc7a32b8dda5ff90f13819ce81e06f6c4e373e9e956) target consumers of the project.

snir911 commented on podvm/README.md:
```diff
@@ -0,0 +1,128 @@
+The builder image packages the cloud-api-adaptor and Kata Containers sources as well as the softwares to build
+the binaries (e.g. *kata-agent* and *agent-protocol-forwarder*) that should be installed in the podvm image.
+
+The builder image is agnostic to cloud providers in the sense that one can be used to buil for multiple providers, however it is
```
typo: "buil"

wainersm replied:
you really read the doc :)

snir911 commented on podvm/README.md:
````diff
@@ -0,0 +1,128 @@
+As an example, to build the builder image for Ubuntu, run:
+
+```
+$ docker build -t podvm_builder -f Dockerfile.podvm_builder .
````
Do you know if the "." is actually needed? at least with podman it works without it

wainersm replied:
yes, it does. At least the docker version (20.10.17) I am using:
```
$ docker build -t podvm_builder -f Dockerfile.podvm_builder
"docker build" requires exactly 1 argument.
```

stevenhorsman commented:
> > hi @jtumber-ibm ,
> > I tried to build the s390x image on my x86_64 laptop just like I documented on the podvm/README.md but it failed on the packer step. Did I miss something?
> > Also, I am okay to drop my extract-qcow2.sh in favor of your script. I will just need to ensure it is merged before and I can change the docs.
> 
> hi @wainersm, did you have the s390x components installed? i.e. rust target (s390x-unknown-linux-gnu), qemu (qemu-system-s390x) and gcc (gcc-s390x-linux-gnu).
> 
> Feel free to use my script, I just need a couple other IBMers to test the ibmcloud part of the script but should be able to get that merged soon or if that gets delayed I can create a separate PR for the qcow2 from image bit.

FYI - this is high on my list, so I hope to get to it this morning.

snir911 commented on podvm/README.md:
```diff
@@ -0,0 +1,128 @@
+|KATA\_SRC\_BRANCH | CCv0 | The Kata Containers repository branch |
+|GO\_VERSION | 1.18.7 | Go version |
+|PROTOC\_VERSION | 3.11.4 | [Protobuf](https://github.com/protocolbuffers/protobuf) version |
+|RUST\_VERSION | 1.66.0 | Rust version |
```
I think we will be able to drop this in favor of allowing to pass CAA_SRC/BRANCH and KATA_SRC/BRANCH to the build containers (and there override the repos) BUT we currently don't have this option so i guess it's fine :)

snir911 commented on podvm/README.md:
````diff
@@ -0,0 +1,128 @@
+	-f Dockerfile.podvm_builder .
+```
+
+## Building the podvm container image
````
how about just "Building the podvm qcow2 image", below you can mentioned it's wrapped in a container image

snir911 commented on podvm/README.md:
```diff
@@ -0,0 +1,128 @@
+   software versions (e.g. Golang, Rust) as much as possible.
+1. Create the podvm image dockerfile by copying `Dockerfile.podvm` to `Dockerfile.podvm.DISTRO` and adjusting the file
+   properly likewise. In particular, the *PODVM_DISTRO* and *BUILDER_IMG* arguments should be changed.
+1. Create the packer directory (`mkdir qcow2/DISTRO`) where the `qemu-DISTRO.pkr.hcl` and `variables.pkr.hcl` files should be placed. Also on this step you can also use an existing configuration (e.g. `qcow2/ubuntu`) as a template. Ensure that common scripts and files like the `qcow2/misc-settings.sh` are adjusted to support the DISTRO if needed.
```
Makefile should be modified as well

wainersm replied:
True. thanks, I will add that.

snir911 reviewed (commented):
Thanks Wainer! it is super useful!, i added some comments, also i think worth pointing to it from the install instructions

snir911 commented:
when it comes to RHEL build you just need to run it from a subscribed system (unless pkgs needs to be installed in the packer vm then you'll need to use libguestfs to install it e.g. open-vm-tools for vsphere) cc @whitebrandy to confirm

stevenhorsman reviewed (approved):
This is really helpful, thanks! I wonder if we should replace some of the sections in the [libvirt doc](https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/libvirt/README.md) (and others later probably) with a link to this now?
I've tried both methods and left a bit of feedback. I didn't test extract-qcow though if it is likely to be replaced with JT's code.

jtumber-ibm commented:
@wainersm [podvm/hack/download-image.sh](https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/podvm/hack/download-image.sh) merged

wainersm commented:
> @wainersm [podvm/hack/download-image.sh](https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/podvm/hack/download-image.sh) merged

Cool! I will drop my script and adjust the docs. thanks @jtumber-ibm !

wainersm commented on podvm/README.md:
```diff
@@ -0,0 +1,128 @@
+
+# How to build on locally
+
+In order to build locally you must [packer](https://www.packer.io/) and QEMU installed in your host. Apart from those, it requires the source trees and software mentioned on the [developer's guide](docs/DEVELOPMENT.md).
```
Yeah, we got dependencies to build CAA binaries, kata containers as well as the image itself.

qemu-utils, cloud-utils and qemu-kvm are likely needed by packer. Rust, musl and the other software for the kata binaries.

I will try to re-phase that to capture the mess this whole thing is :)

wainersm commented on docs/consuming-prebuilt-podvm-images.md:
```diff
@@ -0,0 +1,14 @@
+# Consuming pre-built podvm images
+
+The project has automation in place to build container images that contain the qcow2 file for the various Linux distributions and cloud providers supported. Those images are re-built after each pull request are merged into this repository.
```
you are right. I got it wrong that we were building the images for every pull request. I fixed that information on the version of this PR that I just pushed.

wainersm commented:
With the latest updates on this PR I addressed all comments made by @snir911 , @stevenhorsman and @bpradipt . I also adjusted to reference the download-image.sh script created by @jtumber-ibm 

Pls let me know if it is good to get merged.

bpradipt commented on docs/consuming-prebuilt-podvm-images.md:
````diff
@@ -0,0 +1,14 @@
+$ ./hack/download-image.sh quay.io/confidential-containers/podvm-libvirt-ubuntu . -o podvm.qcow2
+```
+
+In case your workload images are pulled from a private registry then you need to provide the authentication file by either [installing along with the cloud-api-adaptor deployment](registries-authentication.md#deploy-authentication-file-along-with-cloud-api-adaptor-deployment) or [statically embeddeding in the podvm](registries-authentication.md#statically-embed-authentication-file-in-podvm-image) ...
````
nit: "statically embeddeding in the podvm" => "statically **embedding** in the podvm **image**"

nit: "so find detailed instructions on" => "so find detailed instructions **in**"

bpradipt commented on podvm/README.md:
```diff
@@ -0,0 +1,134 @@
+# Introduction
+
+This directory contains the sources to build the podvm image (qcow2 file) for various Linux distributions and cloud providers. So use
+the instructions on the next sections if you need to build your own image with changes to meet your requirements. Otherwise you can
```
nit: "So use the instructions on the next sections" => "So use the instructions **in** the next sections"

bpradipt commented on podvm/README.md:
```diff
@@ -0,0 +1,134 @@
+
+# How to build locally
+
+In order to build locally it requires the source trees and softwares mentioned on the [developer's guide](../docs/DEVELOPMENT.md) to build this project binaries. It will also need [packer](https://www.packer.io/) (to build the qcow2), [rust](https://www.rust-lang.org/tools/install) (to build the Kata Containers's agent), as well as the following packages:
```
nit: "on the [developer's guide]" => "in the [developer's guide]"

bpradipt commented on podvm/README.md:
```diff
@@ -0,0 +1,134 @@
+the binaries (e.g. *kata-agent* and *agent-protocol-forwarder*) that should be installed in the podvm image.
+
+The builder image is agnostic to cloud providers in the sense that one can be used to build for multiple providers, however it is
+dependent of the Linux distribution the image is built for. Therefore, in this directory you will find dockerfiles to each supported distributions, which are currently Ubuntu 20.04 ([Dockerfile.podvm_builder](./Dockerfile.podvm_builder)), CentOS Stream 8 ([Dockerfile.podvm_builder.centos](./Dockerfile.podvm_builder.centos)), and RHEL 8.7 ([Dockerfile.podvm_builder.rhel](./Dockerfile.podvm_builder ...
```
nit: "of the Linux" => "on the Linux"
nit: "in this directory you will find dockerfiles to each supported distributions" => "in this directory you will find dockerfiles **for** each supported distributions"

bpradipt commented on podvm/README.md:
```diff
@@ -0,0 +1,134 @@
+|PROTOC\_VERSION | 3.11.4 | [Protobuf](https://github.com/protocolbuffers/protobuf) version |
+|RUST\_VERSION | 1.66.0 | Rust version |
+
+As it can be noted on the table above the cloud-api-adaptor repository is cloned within the builder image, so rather than
```
nit: "on the table above" => "in the table above"

bpradipt commented on podvm/README.md:
```diff
@@ -0,0 +1,134 @@
+|RUST\_VERSION | 1.66.0 | Rust version |
+
+As it can be noted on the table above the cloud-api-adaptor repository is cloned within the builder image, so rather than
+copying the local source tree, it will be used the upstream source. But if you want to test local changes then you should:
```
nit: "it will be used the upstream source" => "it will be **using** the upstream source"

bpradipt commented on podvm/README.md:
```diff
@@ -0,0 +1,134 @@
+As it can be noted on the table above the cloud-api-adaptor repository is cloned within the builder image, so rather than
+copying the local source tree, it will be used the upstream source. But if you want to test local changes then you should:
+
+* Push the changes to elsewhere (e.g. https://github.com/$USER/cloud-api-adaptor/tree/my-changes-in-a-branch).
```
nit: "to elsewhere" => "to your fork in github"

bpradipt commented on podvm/README.md:
```diff
@@ -0,0 +1,134 @@
+
+The process to build the podvm container image will effectively compile binaries and create the qcow2 file.
+
+In order to build the podvm image there should be used the corresponding dockerfile of the Linux distro which the builder image (see the [previous step](#building-a-builder-image)) was built. For example, if the builder image was built with *Dockerfile.podvm_builder.DISTRO* then you should use the *Dockerfile.podvm.DISTRO* to build the podvm image.
```
"In order to build the podvm image there should be used the corresponding dockerfile of the Linux distro which the builder image was built" => "In order to build the podvm image you should be using the corresponding dockerfile of the Linux distro for which the builder image was built"

bpradipt commented on podvm/README.md:
```diff
@@ -0,0 +1,134 @@
+
+In order to build the podvm image there should be used the corresponding dockerfile of the Linux distro which the builder image (see the [previous step](#building-a-builder-image)) was built. For example, if the builder image was built with *Dockerfile.podvm_builder.DISTRO* then you should use the *Dockerfile.podvm.DISTRO* to build the podvm image.
+
+The builder image has to be indicated via `BUILDER_IMG` build argument to docker. Apart from that, the `CLOUD_PROVIDER` argument is mandatory and specify the cloud provider that binaries should be built for. Once again, one builder image can be used to build for multiple
```
nit: "and specify the cloud provider " => "and **specifies** the cloud provider"

bpradipt commented on podvm/README.md:
```diff
@@ -0,0 +1,134 @@
+
+# How to add support for a new Linux distribution
+
+In order to add a new Linux distribution essentially it is needed to create some dockerfile files and
```
nit: "dockerfile files" => "dockerfiles"

wainersm commented:
Addressed the latest comments from @bpradipt on this update.

bpradipt reviewed (approved):
/lgtm
Thanks @wainersm
